lib/shared_file_lock.js

```diff
@@ -0,0 +1,70 @@
+import assert from 'assert';
+import Debug from 'debug';
+import fs from 'fs-ext';
+import mzfs from 'mz/fs';
+import Promise from 'promise';
+
+let debug = Debug('docker-worker:lib:shared_file_lock');
+
+export default class SharedFileLock {
+  /* This acts as a semaphore which locks a file if one or more locks 
+   * are acquired on this object until they are released.
+   * 
+   * @param lockFile Name of file to be locked. Must already exist and not be exclusively locked.
+   */
+  constructor(lockFile) {
+    this.count = 0;
+    this.lockFile = lockFile;
+    this.locked = false;
+  }
+
+  //acquires a lock, at >=1 locks it will flock the lockfile
+  async acquire() {
+    if(!this.lockFd) {
+      this.lockFd = await mzfs.open(this.lockFile, 'r');
```

Comment:
I actually liked that you were giving it a file handle... because even if the file is unlinked (deleted) the file handle will still exist, so you should always be able to at least attempt to lock and unlock...

Now we have another possible error case where container and session could race to delete the file... cause this to throw and exception... well, okay I'm not sure deleting the file inside the container deletes it outside. But locking the file handle that was used to create the file was quiet brilliant design in the first place...

I just wanted you to document, that the constructor took a file handle  and not a file name.. 

Note, another option is to move `await mzfs.open(this.lockFile, 'r');` into a setup method, and call that setup method in place of `await fs.writeFile(this.lock);` in `lib/features/interactive.js`..

I think this is mostly nit picking... but with locks it's important to be careful and not allow "interesting" things to happen. Too often it leads to hard to debug issues and deadlocks. Particularly, if someday we were to make another plugin/feature that wanted to lock.